library(data.table)  # for fread
library(rtracklayer)  # for readGFF
library(plyr)   # for rename

dyn.load(paste("RPluMA", .Platform$dynlib.ext, sep=""))
source("RPluMA.R")

input <- function(inputfile) {
  parameters <<- read.table(inputfile, as.is=T);
  rownames(parameters) <<- parameters[,1];
    pfix = prefix()
  if (length(pfix) != 0) {
     pfix <- paste(pfix, "/", sep="")
  }

  # Load transcript GTF tables for merged assembled transcripts generated by StringTie
  read_annot <<- readGFF(file.path(paste(pfix, parameters["gtffile", 2], sep="/")))

  # Only transcripts with gene names
  read_annot <<- read_annot[which(!is.na(read_annot$gene_name)), ]

  # Load gene expression count matrices generated by tuxedo2 
  # .cvs files collated by prepDE.py
  count_mat <<- fread(file.path(paste(pfix, parameters["csvfile", 2], sep="/")))
  count_mat <<- rename(count_mat, c("V1"="gene_id"))  # rename for convenience
}

run <- function() {
  # Gene annotation table matching the count matrix
  gene_annot <<- data.frame(gene_id=count_mat$gene_id)

  # Map gene IDs to the merged GTF file specifiying trancsripts detected across samples
  idx = match(gene_annot$gene_id, read_annot$gene_id)

  message("Gene IDs without match: ", sum(is.na(idx)))

  # Transfer to annotation table
  gene_annot$gene_name <<- read_annot$gene_name[idx]
  gene_annot$ref_gene_id <<- read_annot$ref_gene_id[idx]

  # gene_name = read_annot$gene_name[idx]
  message("Transcripts without gene names: ", sum(is.na(gene_annot$gene_name)))
}

output <- function(outputfile) {
  write.csv(gene_annot,
	file.path(outputfile),
	row.names=FALSE,
	quote=FALSE)
}
